### PR TITLE
Add new LXC: Spoolman

### PR DESCRIPTION
**Pull request by MickLesk**
**Spoolman - Keep track of your inventory of 3D-printer filament spools.**

Spoolman is a self-hosted web service designed to help you efficiently manage your 3D printer filament spools and monitor their usage. It acts as a centralized database that seamlessly integrates with popular 3D printing software like [OctoPrint](https://octoprint.org/) and [Klipper](https://www.klipper3d.org/)/[Moonraker](https://moonraker.readthedocs.io/en/latest/). When connected, it automatically updates spool weights as printing progresses, giving you real-time insights into filament usage.

## Description
- Add new LXC Spoolman for Printer Filament
- Tested & Perfectly installed
- Update Script included & functionally

Request: https://github.com/tteck/Proxmox/discussions/50#discussioncomment-9750025
Source: https://github.com/Donkie/Spoolman

- [x] New Script (Develop a new script or set of scripts that are fully functional and thoroughly tested)
- [x] I have performed a self-review of my code, adhering to established codebase patterns and conventions.
- [x] This change requires a documentation update

image

### CODE REVIEW

File: ct/spoolman.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env bash
+source <(curl -s https://raw.githubusercontent.com/tteck/Proxmox/main/misc/build.func)
+# Copyright (c) 2021-2024 tteck
+# Author: MickLesk (Canbiz)
```
Comment (tteck) [inline ct/spoolman.sh]: Co-Author

File: ct/spoolman.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  SD=""
+  NS=""
+  MAC=""
+  VLAN=""
+  SSH="no"
+  VERB="no"
+  echo_default
+}
+
+function update_script() {
```
Comment (tteck) [inline ct/spoolman.sh:56]: Use conditional update. Keep `spoolman_bak` as a precaution until next update. Corrected `msg_ok "Cleaning up Successfully!"` Refactored.
```
function update_script() {
header_info
if [[ ! -d /opt/spoolman ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
if (( $(df /boot | awk 'NR==2{gsub("%","",$5); print $5}') > 80 )); then
  read -r -p "Warning: Storage is dangerously low, continue anyway? <y/N> " prompt
  [[ ${prompt,,} =~ ^(y|yes)$ ]] || exit
fi
RELEASE=$(wget -q https://github.com/Donkie/Spoolman/releases/latest -O - | grep "title>Release" | cut -d " " -f 4)
if [[ ! -f /opt/${APP}_version.txt ]] || [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]]; then

  msg_info "Stopping ${APP} Service"
  systemctl stop spoolman
  msg_ok "Stopped ${APP} Service"

  msg_info "Updating ${APP} to ${RELEASE}"
  cd /opt
  rm -rf spoolman_bak
  mv spoolman spoolman_bak
  wget -q https://github.com/Donkie/Spoolman/releases/download/${RELEASE}/spoolman.zip 
  unzip -q spoolman.zip -d spoolman 
  cd spoolman
  pip3 install -r requirements.txt >/dev/null 2>&1
  cp .env.example .env
  echo "${RELEASE}" >/opt/${APP}_version.txt
  msg_ok "Updated ${APP} to ${RELEASE}"

  msg_info "Starting ${APP} Service"
  systemctl start spoolman
  msg_ok "Started ${APP} Service"

  msg_info "Cleaning up"
  rm -rf /opt/spoolman.zip
  msg_ok "Cleaned"

  msg_ok "Updated Successfully!\n"
else
  msg_ok "No update required. ${APP} is already at ${RELEASE}"
fi
exit
}
```

File: install/spoolman-install.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+source /dev/stdin <<<"$FUNCTIONS_FILE_PATH"
+color
+verb_ip6
+catch_errors
+setting_up_container
+network_check
+update_os
+
+msg_info "Installing Dependencies"
```
Comment (tteck) [inline install/spoolman-install.sh:18]: Don't install unnecessary dependencies. Removed `--no-install-recommends`, `unzip`, `jq` and `git`
```
$STD apt-get install -y \
  build-essential \
  curl \
  sudo \
  make \
  libpq-dev \
  gpg \
  ca-certificates \
  mc
```

File: install/spoolman-install.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  jq \
+  git \
+  make \
+  libpq-dev \
+  gnupg \
+  ca-certificates \
+  mc
+msg_ok "Installed Dependencies"
+
+msg_info "Installing Python3"
```
Comment (tteck) [inline install/spoolman-install.sh:30]: `python3` already installed. `venv` is not needed
```
$STD apt-get install -y \
  python3-dev \
  python3-setuptools \
  python3-wheel \
  python3-pip
```

File: install/spoolman-install.sh [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+$STD apt-get install -y \
+  python3 \
+  python3-dev \
+  python3-setuptools \
+  python3-wheel \
+  python3-pip \
+  python3-venv
+msg_ok "Installed Python3"
+
+msg_info "Installing Spoolman"
```
Comment (tteck) [inline install/spoolman-install.sh:38]: Install without creating a virtual environment. The `/scripts` directory is not needed
```
RELEASE=$(wget -q https://github.com/Donkie/Spoolman/releases/latest -O - | grep "title>Release" | cut -d " " -f 4)
cd /opt
wget -q https://github.com/Donkie/Spoolman/releases/download/$RELEASE/spoolman.zip
unzip -q spoolman.zip -d spoolman
rm -rf spoolman.zip
cd spoolman
$STD pip3 install -r requirements.txt
cp .env.example .env
echo "${RELEASE}" >/opt/${APPLICATION}_version.txt
```

File: install/spoolman-install.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+echo "${RELEASE}" >/opt/${APPLICATION}_version.txt
+cd spoolman
+$STD python3 -m venv .venv
+$STD source .venv/bin/activate
+$STD pip3 install -r requirements.txt
+cp .env.example .env
+chmod +x scripts/*.sh
+msg_ok "Installed Spoolman"
+
+msg_info "Creating Service"
```
Comment (tteck) [inline install/spoolman-install.sh:51]: Call main app for start. Not blindly execute `start.sh`
```
cat <<EOF >/etc/systemd/system/spoolman.service
[Unit]
Description=Spoolman
After=network.target
[Service]
Type=simple
WorkingDirectory=/opt/spoolman
EnvironmentFile=/opt/spoolman/.env
ExecStart=uvicorn spoolman.main:app --host 0.0.0.0 --port 7912
Restart=always
User=root
[Install]
WantedBy=multi-user.target
EOF
```

File: ct/spoolman.sh (hunk 1 above)
Comment (MickLesk) [reply ct/spoolman.sh]: my mistake, I forgot to update my template :-)

File: ct/spoolman.sh (hunk 2 above)
Comment (MickLesk) [reply ct/spoolman.sh:56]: Allright, Done 👍🏻

File: install/spoolman-install.sh (hunk 3 above)
Comment (MickLesk) [reply install/spoolman-install.sh:18]: unzip is needed, mc dont have unzip, the others have been removed

"mc (Midnight Commander) doesn't have a direct dependency on unzip in Debian.
mc has built-in functionality to unpack and navigate ZIP archives.
unzip is a separate tool for extracting ZIP archives with more features."

Comment (tteck) [reply install/spoolman-install.sh:18]: It does when `--no-install-recommends` is removed

File: install/spoolman-install.sh [hunk 7]
```diff
@@ -73,3 +73,5 @@ msg_info "Cleaning up"
 $STD apt-get -y autoremove
 $STD apt-get -y autoclean
 msg_ok "Cleaned"
+echo -e "${APP} Setup should be reachable by going to the following URL.
```
Comment (tteck) [inline install/spoolman-install.sh]: That is on the `spoolman.sh`

**Comment by tteck** [comment]
Have you performed all tests? Is she ready to merge?

**Comment by MickLesk** [comment]
Tested and working. (After Commits) 




> Have you performed all tests? Is she ready to merge?


Give me 5 minutes, I'll deploy it on my test proxmox.

**Comment by MickLesk** [comment]
Ready to merge. Works great

image

**Comment by hgelpke** [comment]
firefox_EiaRc1VcZo

Same issue today. Both in fresh install and upgrade

**Comment by MickLesk** [comment]
Run without verbose mode and send Textfile again

**Pull request merged**